### ISSUE TITLE
Consider allowing provider partition on ProviderConfig

### ISSUE THREAD

**Issue by hasheddan** [crossplane/crossplane#2411]
### What problem are you facing?

As users continue to adopt Crossplane, installed Providers have become responsible for an increasing number of managed resources. As the number of managed resources climbs, Providers are more likely to run into resource constraints (i.e. CPU / memory), as well as controller rate-limiting, and any rate-limiting imposed by the external API. To address the first issue, infrastructure admins can always allocate more and more resources to the Provider `Deployment`, but the latter two issues are more complicated.

External rate limiting generally is based on factors such as the credentials being used to authenticate to the API, or characteristics of the machine making the request (e.g. IP address). Controller rate-limiting can be adjusted, but only at the per-controller and global level. We [have explored](https://github.com/crossplane/crossplane-runtime/issues/253#issue-830211294) exposing the polling interval for managed resources on their referenced `ProviderConfig`, which would make the drift detection frequency (not the rate-limiting) configurable, and we are [in the process](https://github.com/crossplane/provider-aws/pull/741) of exposing the global drift detection interval at the Provider `Pod` level. In practice, however, setting polling interval on the `ProviderConfig` as Providers currently are deployed today could become a rather complicated system where managed resources have many different polling intervals, but are all governed by the same global and per-controller rate-limiting (backoff) restrictions.

In addition to performance and latency implications, as Crossplane deployments scale within organizations, various permission models are required to satisfy the constraints of each organization. Having a single Provider `Pod` that reconciles all managed resources can make these constraints difficult to operate within, especially if consumers are using techniques such as IRSA or some other form of workload identity.

### How could Crossplane help solve your problem?

As dictated by [Provider conformance testing](https://github.com/crossplane/conformance/blob/fce75e1aed26a057d7fa166ba56787b8e9182f97/provider/provider_test.go#L197), any conformant Crossplane Provider must supply and support the usage of a `ProviderConfig`. As evidenced by some of the thinking around exposing the polling interval on `ProviderConfig`, as well as being the fact that a `ProviderConfig` is a set of, well, configuration for a Provider, it makes sense to use them as a unit of partitioning for Provider deployments.

This could be supported as an opt-in feature when installing a Provider, and Providers could even indicate whether they support the feature or not in their package manifest (i.e. `crossplane.yaml` -- this could be implicitly denoted by setting a Crossplane constraint to >= the version where this functionality was introduced). When a consumer opts-in to Provider partitioning, I imagine the functionality providing the following experience:

1. In a Provider's `crossplane.yaml`, there is an option to specify the group and version for the `ProviderConfig`. If not specified, Crossplane will assume it is in the top-level group (name with `provider-` stripped + `crossplane.io`) and `v1beta1` (for now). Alternatively, we could mandate that these fields be provided in the `crossplane.yaml` as an indication that partitioning is supported.
```yaml
apiVersion: meta.pkg.crossplane.io/v1alpha1
kind: Provider
metadata:
  name: provider-aws
spec:
  controller:
    image: crossplane/provider-aws-controller:v0.19.0
  providerConfig:
    group: aws.crossplane.io
    version: v1beta1
```
2. Package author builds and pushes package.
3. Infra admin installs Provider and specifies `providerConfigPartitioning: true`. Alternatively, this could be specified on a `ControllerConfig`, which would allow us more flexibility with the API as it is `v1alpha1` and is already used to "configure controllers".
```yaml
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: provider-gcp
spec:
  package: crossplane/provider-aws:v0.19.0
  packagePullPolicy: IfNotPresent
  providerConfigPartitioning: true
  revisionActivationPolicy: Automatic
  revisionHistoryLimit: 1
```
4. Package manager goes through normal install process, but, after installing CRDs, revision controller instead starts a new controller that watches for instances of the specified `ProviderConfig` and creates a new `Deployment` for each `ProviderConfig`.
5. Provider controller process must support (i.e. becomes part of conformance) a flag / env var `PC_PARTITION` that specifies the name of the `ProviderConfig` for which it will reconcile managed resources that reference it. This would be passed to a predicate on each controller in the Provider. 

This effort would absolutely require a fairly robust design doc before implementation, especially exploring how each partition could be configured (e.g. it may be necessary to allow setting a different `ServiceAccount` per partition for workload identity scenarios, and it may be desirable to only add something like the [Vault sidecar](https://crossplane.io/docs/v1.2/guides/vault-injection.html) to a subset of partitions). However, the general pattern seems like a viable solution, so I would like to open up feedback and see if this is a desirable path for the community.

**Comment by muvaf** [comment]
From what I can tell, the constraints users have been reporting are around CPU/memory limitation of the controller and API rate-limiting of the provider. As mentioned, CPU/memory problem is alleviated by increasing the allocated memory through `ControllerConfig`. But I'm not sure if partitioning will help external API rate limiting. For example, AWS imposes call limit [**per account** and **per region**](https://aws.amazon.com/premiumsupport/knowledge-center/elastic-beanstalk-api-throttling-errors/), which means even if two different `Pod`s on different `Node`s make the calls, if they do them on behalf of the same account, it won't be different than single `Pod`.

**Comment by hasheddan** [comment]
@muvaf that is true for some providers (such as AWS), but there is still value in partitioning on the `ProviderConfig` because it allows you to configure which resources get to "consume" the allotment of API requests that can be made. For instance, if you have resources that are being used for non-critical workloads, you could have them use all `ProviderConfig` that was tied to a partition with a very infrequent polling rate, thus reducing the number of total API requests and allowing more critical infrastructure to be monitored more closely. We can't change the way rate limiting works externally, so we should do our best to allow users to work within its bounds.

Also, anecdotally, I have seen more concerns around global rate-limiting in the controllers themselves (rather than the external provider), where a few misconfigured resources could drive up the backoff for other resources. Partitioning could support cases where one team's faulty infrastructure does not have to impact the reconciliation of another team's properly configured and functional infrastructure.  

Also, this is not just tied to performance / rate limiting. It is valuable for authentication scenarios such as IRSA, where you could split the provider into multiple `Pod`s based on `ProviderConfig`, each being able to assume its credentials using the IAM role assigned to the `Pod`'s `ServiceAccount`.

**Comment by muvaf** [comment]
Got it, so the goal is more around giving the option to the users so that they can selectively allocate CPU/memory/rate to different sets of resources.

> It is valuable for authentication scenarios such as IRSA, where you could split the provider into multiple Pods based on ProviderConfig, each being able to assume its credentials using the IAM role assigned to the Pod's ServiceAccount.

I think we discussed that a while ago and partitioning came up but I couldn't find the thread. Yes, I believe partitioning would be a great solution for that problem.

**Comment by mzingarelli** [comment]
We are facing a similar issue as you have outlined but its not performance related, rather a separation of concerns. 
We have very strict policies that are driven with specific roles. We are not allowed to use any other method beside IRSA.
An example of this would be `RDS-Admin-Role`, `KMS-Admin-Role` and so on.

Crossplane only supports a single `ControllerConfig` to inject the annotation `eks.amazonaws.com/role-arn` for a single `provider-aws`.

We have attempted to use 2 `ControllerConfigs` and 2 `Providers` with different names but the problem is that both providers attempt to establish control of the same objects.

I think a solution for this would be to dynamically specify the control of which objects are managed by the providers

**Comment by negz** [comment]
Sorry for missing discussion in the community meeting this morning. Are we really ready to commit to this (major) architectural change in Crossplane v1.5? Should we perhaps move this to a one pager stage for more discussion and keep it on the roadmap project (in the 'In Design') column?

**Comment by muvaf** [comment]
@negz I'm not aware of a decision we made in the community meeting to take on this.

**Comment by hasheddan** [comment]
@negz ah I unintentionally moved this TODO in v1.5 board but just wanted to propose -- I'll write up a one-pager around how this could work / some other options 👍🏻

**Comment by erikgb** [comment]
Disclaimer: I am quite new to Crossplane, so not aware of the inner details, but I just wanted to throw some comments and ideas on the table - from a Crossplane user perspective.

Faced with the issue of pulling provider credentials from Hashicorp Vault following [the docs](https://crossplane.io/docs/v1.5/guides/vault-injection.html), for multiple environments of the same type (same Provider, but different ProviderConfig), I realized that I only had one ControllerConfig to support all ProviderConfigs - since the ControllerConfig is referenced from the Provider. And Crossplane does not allow you to create multiple instances of the same provider (type). I [asked on Slack](https://crossplane.slack.com/archives/C01718T2476/p1636108283133700), and got immediate response from @hasheddan.

My first thought was: Why am I not able to reference a ControllerConfig from a **ProviderConfig**? I would expect that to spin up a dedicated controller (in a separate workload) for the ProviderConfig - responsible for reconciliation of managed resources referencing that ProviderConfig.

That could solve this "partition on ProviderConfig" issue (and other related issues; like mine), but I'll guess there are several challenges with this approach?

**Comment by muvaf** [comment]
@hasheddan FWIW, I don't think this has to be optional on provider level. Today, there is no way that you can make use of a provider without a `ProviderConfig` object anyway. So, for users, the only difference will be that they will see more `Deployment` objects which we can assess whether has some performance implications (I assume we'll need some event filtering based on `providerConfigRef.name` value, which means all controllers will get the event in all cases). But other than that, I think we can go ahead and make it a default and add a flag to Crossplane itself to disable it.

**Comment by erikgb** [comment]
I agree with @muvaf that this could be the new default. If I have understood this correctly, it will only make a difference for users with more than one ProviderConfig for any Provider - and isn't that the use case we want to solve here?

To illustrate my idea, I will show an example on how I have to solve my use case with the current architecture, and how I as a user would like to see this in the future:

Now:
````yaml
---
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: panos
spec:
  package: <panos-provider image>
  controllerConfigRef:
    name: panos
---
apiVersion: pkg.crossplane.io/v1alpha1
kind: ControllerConfig
metadata:
  name: panos
spec:
  metadata:
    annotations:
      vault.hashicorp.com/agent-inject: "true"
      vault.hashicorp.com/role: crossplane
      vault.hashicorp.com/agent-inject-secret-panorama-staging.txt: crossplane/panos/panorama-staging
      vault.hashicorp.com/agent-inject-template-panorama-staging.txt: |
        {{ with secret "crossplane/panos/panorama-staging" }}{{ .Data.data.apiKey }}{{ end }}
      vault.hashicorp.com/agent-inject-secret-panorama-production.txt: crossplane/panos/panorama-production
      vault.hashicorp.com/agent-inject-template-panorama-production.txt: |
        {{ with secret "crossplane/panos/panorama-production" }}{{ .Data.data.apiKey }}{{ end }}
---
apiVersion: panos.crossplane.io/v1alpha1
kind: ProviderConfig
metadata:
  name: panorama-staging
spec:
  hostname: <panorama hostname>
  deviceGroup: <panaorama deviceGroup>
  credentials:
    source: Filesystem
    fs:
      path: /vault/secrets/panorama-staging.txt
---
apiVersion: panos.crossplane.io/v1alpha1
kind: ProviderConfig
metadata:
  name: panorama-production
spec:
  hostname: <panorama hostname>
  deviceGroup: <panaorama deviceGroup>
  credentials:
    source: Filesystem
    fs:
      path: /vault/secrets/panorama-production.txt
````

Future:
````yaml
---
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: panos
spec:
  package: <panos-provider image>
---
apiVersion: panos.crossplane.io/v1alpha1
kind: ProviderConfig
metadata:
  name: panorama-staging
spec:
  hostname: <panorama hostname>
  deviceGroup: <panaorama deviceGroup>
  credentials:
    source: Filesystem
    fs:
      path: /vault/secrets/creds.txt
  controllerConfigRef:
    name: panorama-staging
---
apiVersion: pkg.crossplane.io/v1alpha1
kind: ControllerConfig
metadata:
  name: panorama-staging
spec:
  metadata:
    annotations:
      vault.hashicorp.com/agent-inject: "true"
      vault.hashicorp.com/role: crossplane-staging
      vault.hashicorp.com/agent-inject-secret-creds.txt: crossplane/panos/panorama-staging
      vault.hashicorp.com/agent-inject-template-creds.txt: |
        {{ with secret "crossplane/panos/panorama-staging" }}{{ .Data.data.apiKey }}{{ end }}
---
apiVersion: panos.crossplane.io/v1alpha1
kind: ProviderConfig
metadata:
  name: panorama-production
spec:
  hostname: <panorama hostname>
  deviceGroup: <panaorama deviceGroup>
  credentials:
    source: Filesystem
    fs:
      path: /vault/secrets/creds.txt
  controllerConfigRef:
    name: panorama-staging
---
apiVersion: pkg.crossplane.io/v1alpha1
kind: ControllerConfig
metadata:
  name: panorama-production
spec:
  metadata:
    annotations:
      vault.hashicorp.com/agent-inject: "true"
      vault.hashicorp.com/role: crossplane-production
      vault.hashicorp.com/agent-inject-secret-creds.txt: crossplane/panos/panorama-production
      vault.hashicorp.com/agent-inject-template-creds.txt: |
        {{ with secret "crossplane/panos/panorama-production" }}{{ .Data.data.apiKey }}{{ end }}
````

The main difference being that the `controllerConfigRef` is now set on the `ProviderConfig` - which allows me to configure the Vault agent injector more precisely, and thus less room for (human) errors impacting the security. This could be thightened even more by specifying `serviceAccountName` on the controller configs, and reflect this in the Vault permissions. More in accordance with _principle of least privilege_.

And I think this also will solve the original use case as described by @hasheddan.

**Comment by hasheddan** [comment]
@erikgb thanks for adding your thoughts here! This issue is discussing specifically spinning up new controllers on each `ProviderConfig`, whereas configuring each of those controllers (which I believe we must support as the credentials use cases you describe are also called out as a motivation for this design in the original body) is a follow-on effect of the partitioning. To support this, I have proposed a [Provider Runtime Interface (PRU)](https://github.com/crossplane/crossplane/issues/2671) that would allow for configuring controllers running as `Deployments` (as they do today), but also via other mechanisms (e.g. external functions, etc.). In that model, the configuration mechanism would likely need to be deferred to each PRI implementation, but I definitely agree that selecting the specific configuration should be done via the `ProviderConfig` as you mention. This could be accomplished via references (as you have shown) or label selectors, and that information would be passed to the PRI implementation, which would bring its own "controller config" implementation based on the underlying compute medium.

TLDR; I think a lot of the thoughts you present here are very applicable to #2671 :)

> Today, there is no way that you can make use of a provider without a ProviderConfig object anyway.

@muvaf this is true by convention, but not enforced in any way, and Crossplane is also not knowledgeable of the GVK of the `ProviderConfig` of a given `Provider`, so the package would need to declare it in some form or fashion. I like the idea of making this an alpha feature in Crossplane first, but I think even when it is enabled a `Provider` should indicate that it supports the mode rather than Crossplane just assuming that it does by default.

**Comment by shailsirohi** [comment]
Hi - Let me start by mentioning I am very new to crossplane. So, may have some gaps in my understanding.

This is a similar use case that we are struggling with. We are creating a central multi-tenant control plane where application teams will be onboarded. EKS Namespace as isolation mechanisms between teams. App teams will only be able to add claims and not clustered scope objects and provider config patched at composition level. We are also following IRSA model. To ensure the teams can only take actions allowed to them, app team admin has to provide IAM role in AWS account which is associated with their team and approved by security team. So, each team can have access to multiple accounts and differnet users can have differnet roles in that account. So, we expect number of provider config to be in 100+ as more and more teams onboard. Creating a new pod for each provider config will lead to considerable footprint sprawl and may not be very cost optimal. I would think that using ControllerConfig as base for creating a new deployment along with a new SA and then allowing multiple ProviderConfig to be associated with the same ControllerConfig will be a more flexible solution. This will allow platform operators to choose what's the right amount of horizontal scaling in provider pods is required for their usecase

**Comment by jeanduplessis** [comment]
Based on my understanding, we can close this issue as outdated/no longer needed.
Since this issue was opened, provider families and the DeploymentRuntimeConfig API were introduced.
